### ISSUE TITLE
Hello world example won't run (not /dev/sgx/enclave)

### ISSUE THREAD

**Issue by hansinator** [oscarlab/graphene#2233]
## Description of the problem

I was following the quick start guide for an SGX enabled system and tried running the hello world example.
The pal loader complains that it can not find /dev/sgx/enclave, but I have got the isgx kernel module with /dev/isgx on kernel 5.10.

## Steps to reproduce

Follow quick start guide.

## Expected results

Seeing the words "Hello World" on my console.

## Actual results
```
chris@HP-EliteBook-830-G6:~/graphene/LibOS/shim/test/regression$ SGX=1 ./pal_loader helloworld
Cannot open device /dev/sgx/enclave. Please make sure the Intel SGX kernel module is loaded.
load_enclave() failed with error -2
chris@HP-EliteBook-830-G6:~/graphene/LibOS/shim/test/regression$ ls /dev/isgx 
/dev/isgx
chris@HP-EliteBook-830-G6:~/graphene/LibOS/shim/test/regression$ lsmod | grep sgx
isgx                   53248  2
```

**Comment by mkow** [comment]
You've probably passed wrong path in `ISGX_DRIVER_PATH` when building. Please take a look at https://graphene.readthedocs.io/en/latest/building.html#id1.

**Comment by dnoliver** [comment]
Having the same issue here, I don't have SGX Launch Config support, so I am using the legacy kernel module `isgx`. 

I built and installed the `isgx` module from source:

```shell
root@test-NUC8v5PNK:~# ls /root/linux-sgx-driver/
inker2ext  isgx.mod    isgx.mod.o  License.txt  modules.order   README.md   sgx_asm.h    sgx_encl2.o  sgx_encl.o  sgx_ioctl.c  sgx_main.c  sgx_page_cache.c  sgx_user.h  sgx_util.o  sgx_vma.o
isgx.ko    isgx.mod.c  isgx.o      Makefile     Module.symvers  sgx_arch.h  sgx_encl2.c  sgx_encl.c   sgx.h       sgx_ioctl.o  sgx_main.o  sgx_page_cache.o  sgx_util.c  sgx_vma.c
```

And built graphene as well:

```shell
make ISGX_DRIVER_PATH=/root/linux-sgx-driver/ SGX=1
meson build -Ddirect=disabled -Dsgx=enabled
ninja -C build
ninja -C build install
export PYTHONPATH=$PYTHONPATH:/usr/local/lib/python3.8/site-packages/
cd LibOS/shim/test/regression/
make ISGX_DRIVER_PATH=/root/linux-sgx-driver/ SGX=1
make ISGX_DRIVER_PATH=/root/linux-sgx-driver/ SGX=1 sgx-tokens
```

Then, trying to run it fails:

```shell
root@test-NUC8v5PNK:~/graphene/LibOS/shim/test/regression# graphene-sgx helloworld
error: Cannot open device /dev/sgx/enclave. Please make sure the Intel SGX kernel module is loaded.
error: load_enclave() failed with error -2
```

I can run sample apps with the Intel SGX SDK and with the OpenEnclave SDK with that built kernel module

Any hints?

**Comment by triplewings** [comment]
> Having the same issue here, I don't have SGX Launch Config support, so I am using the legacy kernel module `isgx`.
> 
> I built and installed the `isgx` module from source:
> 
> ```shell
> root@test-NUC8v5PNK:~# ls /root/linux-sgx-driver/
> inker2ext  isgx.mod    isgx.mod.o  License.txt  modules.order   README.md   sgx_asm.h    sgx_encl2.o  sgx_encl.o  sgx_ioctl.c  sgx_main.c  sgx_page_cache.c  sgx_user.h  sgx_util.o  sgx_vma.o
> isgx.ko    isgx.mod.c  isgx.o      Makefile     Module.symvers  sgx_arch.h  sgx_encl2.c  sgx_encl.c   sgx.h       sgx_ioctl.o  sgx_main.o  sgx_page_cache.o  sgx_util.c  sgx_vma.c
> ```
> 
> And built graphene as well:
> 
> ```shell
> make ISGX_DRIVER_PATH=/root/linux-sgx-driver/ SGX=1
> meson build -Ddirect=disabled -Dsgx=enabled
> ninja -C build
> ninja -C build install
> export PYTHONPATH=$PYTHONPATH:/usr/local/lib/python3.8/site-packages/
> cd LibOS/shim/test/regression/
> make ISGX_DRIVER_PATH=/root/linux-sgx-driver/ SGX=1
> make ISGX_DRIVER_PATH=/root/linux-sgx-driver/ SGX=1 sgx-tokens
> ```
> 
> Then, trying to run it fails:
> 
> ```shell
> root@test-NUC8v5PNK:~/graphene/LibOS/shim/test/regression# graphene-sgx helloworld
> error: Cannot open device /dev/sgx/enclave. Please make sure the Intel SGX kernel module is loaded.
> error: load_enclave() failed with error -2
> ```
> 
> I can run sample apps with the Intel SGX SDK and with the OpenEnclave SDK with that built kernel module
> 
> Any hints?

Having exactly the same issue, need help

**Comment by dimakuv** [comment]
@dnoliver @triplewings Can you copy-paste here the autogenerated `Pal/src/host/Linux-SGX/gsgx.h` file? In particular, the last couple lines must contain `ISGX_FILE` and `SGX_DCAP` macros. It looks like something got broken in builds that use `/dev/isgx`.

The steps presented by @dnoliver look correct. I am surprised Graphene incorrectly detects `/dev/sgx/enclave`.

**Comment by dnoliver** [comment]
Here is my `Pal/src/host/Linux-SGX/gsgx.h`

```shell
/* SPDX-License-Identifier: LGPL-3.0-or-later */
/* (C) Copyright 2020 Intel Corporation
 *                    Dmitrii Kuvaiskii <dmitrii.kuvaiskii@intel.com>
 */

#ifndef __ARCH_GSGX_H__
#define __ARCH_GSGX_H__

#ifndef __packed
#define __packed __attribute__((packed))
#endif

#include <linux/stddef.h>
#include <linux/types.h>

#include "/root/graphene/Pal/src/host/Linux-SGX/sgx_in_kernel.h"

#define GSGX_FILE  "/dev/gsgx"

/* Graphene needs the below subset of SGX instructions' return values */
#ifndef SGX_INVALID_SIG_STRUCT
#define SGX_INVALID_SIG_STRUCT  1
#endif

#ifndef SGX_INVALID_ATTRIBUTE
#define SGX_INVALID_ATTRIBUTE   2
#endif

#ifndef SGX_INVALID_MEASUREMENT
#define SGX_INVALID_MEASUREMENT 4
#endif

#ifndef SGX_INVALID_SIGNATURE
#define SGX_INVALID_SIGNATURE   8
#endif

#ifndef SGX_INVALID_EINITTOKEN
#define SGX_INVALID_EINITTOKEN  16
#endif

#ifndef SGX_INVALID_CPUSVN
#define SGX_INVALID_CPUSVN      32
#endif

/* SGX_INVALID_LICENSE was renamed to SGX_INVALID_EINITTOKEN in SGX driver 2.1:
 *   https://github.com/intel/linux-sgx-driver/commit/a7997dafe184d7d527683d8d46c4066db205758d */
#ifndef SGX_INVALID_LICENSE
#define SGX_INVALID_LICENSE     SGX_INVALID_EINITTOKEN
#endif

#define ISGX_FILE "/dev/sgx/enclave"

#define SGX_DCAP 1

#endif /* __ARCH_GSGX_H__ */
```

**Comment by dimakuv** [comment]
> #include "/root/graphene/Pal/src/host/Linux-SGX/sgx_in_kernel.h"

This line is not good. Your `ISGX_DRIVER_PATH` wasn't used at all.

I have a suspicion that you first built incorrectly (maybe with `ISGX_DRIVER_PATH=""`) and then tried to build again. If this is the case, you need to do `make distclean && make distclean SGX=1`. And then build Graphene again, this time from scratch.

**Comment by dnoliver** [comment]
It looks like doing `export ISGX_DRIVER_PATH=/root/linux-sgx-driver/` and then building solve this issue.
I guess the env vars are not being passed correctly to the child make targets?

After building with the exported `ISGX_DRIVER_PATH`, I see this error:

```shell
root@test-NUC8v5PNK:~/graphene/LibOS/shim/test/regression# graphene-sgx helloworld
error: Using insecure argv source. Graphene will continue application execution, but this configuration must not be used in production!
error: {RD,WR}{FS,GS}BASE instructions are not permitted on this platform. Please check the instructions under "Building with SGX support" from Graphene documentation.
error: *** Unexpected exception occurred inside PAL at RIP = +0x0000b7dc! ***
error: (SGX HW reported AEX vector 0x6 with exit_type = 0x3)
error: rax: 0x0ff5e010 rcx: 0x0b68dc60 rdx: 0x00000000 rbx: 0x00000000
rsp: 0x0ff5cc68 rbp: 0x0b63a040 rsi: 0x00000000 rdi: 0x00000000
r8 : 0x0b5f2120 r9 : 0x00000000 r10: 0x0f68a910 r11: 0x0b65cffc
r12: 0x0ff5cd28 r13: 0x00000001 r14: 0x0b5b24d0 r15: 0x0f684c65
rflags: 0x00010246 rip: 0x0b64d7dc
```

I thought the `gsgx` module was there for the `FSGSBASE` support. I have it loaded:

```shell
root@test-NUC8v5PNK:~/graphene/LibOS/shim/test/regression# lsmod | grep gsgx
gsgx                   16384  0
```

Is that being ignored as well?

**Comment by dimakuv** [comment]
What is the Linux version you have? `uname -a`

**Comment by dnoliver** [comment]
Ubuntu 2020 with 5.8 HWE kernel:

```shell
Linux test-NUC8v5PNK 5.8.0-50-generic #56~20.04.1-Ubuntu SMP Mon Apr 12 21:46:35 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by dimakuv** [comment]
Linux 5.8 is a "bad" kernel for Graphene.

There was a period starting from Linux 5.4 where the kernel proactively disables FSGSBASE if it was set by an insecure kernel module like our `/dev/gsgx`. In Linux 5.9, the FSGSBASE functionality was merged in, and thus there is no need for `/dev/gsgx` at all.

So what happens is that your Linux 5.8 detects that `/dev/gsgx` toggles the FSGSBASE feature bit, and then proactively disables it again. You can google "linux fsgsbase" and read about why this happens, you'll have a lot of fun :)

I would recommend you to upgrade your kernel to at least Linux 5.9; this is not too hard to do. You basically don't have another option.

**Comment by dnoliver** [comment]
Installed Kernel 5.9 following instructions from https://ubuntuhandbook.org/index.php/2020/10/linux-kernel-5-9-released/
Checked for FSGSBASE support with `cat /proc/cpuinfo | grep fsgsbase` with success, so stopped using the `gsgx` driver.
Rebuilt and reinstalled `isgx` kernel module with new headers
Restarted `aesmd.service` with new `isgx` kernel module
Cleaned and rebuilt graphene, and now my `helloworld` is working!

```shell
root@test-NUC8v5PNK:~/graphene/LibOS/shim/test/regression# graphene-sgx helloworld
error: Using insecure argv source. Graphene will continue application execution, but this configuration must not be used in production!
Hello world
```

AFAIK, Ubuntu 20.04 use the Kernel 5.4 by default, and goes up to Kernel 5.8 with the HWE package.
So this makes the current Ubuntu 20.04 a "bad" distro for Graphene? 😀 

Thanks for your help!

**Comment by dimakuv** [comment]
> So this makes the current Ubuntu 20.04 a "bad" distro for Graphene? 😀

Well, yes, kinda. You must install a proper Linux kernel for it...

**Comment by dnoliver** [comment]
@hansinator proper cleaning (`make distclean && make distclean SGX=1`) seems to solve this issue as noted in https://github.com/oscarlab/graphene/issues/2233#issuecomment-832736219. 
Doing `export ISGX_DRIVER_PATH=/path/to/linux-sgx-driver` also seems to work. 

And then, having the right kernel version > 5.9 is also a must (I guess < 5.4 with `gsgx` should also work). 

Thanks!

**Comment by jtakeshi-nd** [comment]
I am also having this issue. System is Ubuntu 18.04, with kernel 5.11.0-051100-generic (so the driver is in-kernel). The fix of setting `export ISGX_DRIVER_PATH=""` before building did not work. My gsgx.h has `#define ISGX_FILE "/dev/sgx/enclave"`, but the only files in /dev related to SGX are isgx, sgx_enclave, and sgx_provision.

Upon running the GCC example, the following error occurs:
`graphene-sgx gcc test_files/helloworld.c -o test_files/hello
error: Cannot open device /dev/sgx/enclave. Please make sure the Intel SGX kernel module is loaded.
error: load_enclave() failed with error -2`

This is somewhat confusing, as I managed to successfully build and run SGX/Graphene on a similar or highly identical machine. (Once my colleague replies with the attributes of the other machine, I will edit this to verify that the kernel is the same.)

**Comment by dnoliver** [comment]
From the official documentation of the Kernel 5.11 about SGX (https://www.kernel.org/doc/html/v5.11/x86/sgx.html#enclave-build-functions) the dev nodes are `/dev/sgx_enclave` and `/dev/sgx_provision`.

The DCAP driver mention this discrepancy in https://github.com/intel/SGXDataCenterAttestationPrimitives/blob/94e327a8235bc7c85353dc5e0a6dc1564e6c96c6/driver/linux/README.kernel.md#udev-rules. The proposed solution is to add the udev rules to symlink those dev nodes to the expected ones (rules documented in the link shared)

I *guess* Graphene should try to track the latest in-kernel version? I hope that works for machines with SGX 1 and SGX 2!

**Comment by mkow** [comment]
That's right, the in-kernel driver supports only platforms with FLC (but that doesn't mean SGX2, all these features are independent and to make it even more annoying, you can't check which ones are supported on ARK).

**Comment by dimakuv** [comment]
@jtakeshi-nd @dnoliver I created a PR https://github.com/oscarlab/graphene/pull/2387. We decided to make Linux 5.11+ the default choice for Graphene. With this PR, Graphene will use `/dev/sgx_enclave` by default (if `ISGX_DRIVER_PATH=""`). I'll be grateful if you could test this PR.

**Comment by jtakeshi-nd** [comment]
@dimakuv Thanks! I have tested this on 5.11. Initially, there were errors relating to enclave initialization or loading or freeing pages. I tried adding a group and udev rules as suggested in the link from @dnoliver; this did not work so I removed the group and rules and rebooted. After a reboot, running now works - there is probably a moral somewhere here.

I have tested the curl and GCC tests with this PR, and both pass. Thank you very much for the great level of support that you have provided!
